### ISSUE TITLE
LoRA training crashes on first backward pass with Qwen3.5-9B on M5 Max (applegpu_g17s)

### ISSUE THREAD

**Issue by sivavenkatay** [ml-explore/mlx-lm#1206]
Environment:

macOS, Apple M5 Max (applegpu_g17s), 36GB unified memory
mlx 0.31.2, mlx-lm 0.31.3, mlx-metal 0.31.2
Model: mlx-community/Qwen3.5-9B-4bit
What happens:
Validation (forward pass) completes successfully. The first training iteration crashes immediately with:

[METAL] Command buffer execution failed: Insufficient Memory
(00000008:kIOGPUCommandBufferCallbackErrorOutOfMemory)
This happens regardless of batch size (1 or 2), sequence length (2048–8192), number of LoRA layers (4–16), or whether --grad-checkpoint is set. Total system memory usage is low at the time of crash.

Workaround:
Switching to mlx-community/Qwen3-8B-4bit (same architecture family, previous generation) trains successfully with identical settings. Suggests the issue is specific to Qwen3.5's architecture changes in this mlx version.

Reproduce:

mlx_lm lora \
  --model mlx-community/Qwen3.5-9B-4bit \
  --train --data data/ \
  --batch-size 1 --num-layers 4 \
  --max-seq-length 2048 \
  --grad-checkpoint --val-batches 0

**Comment by lawcontinue** [comment]
This is a different failure mode from the descriptor-count leak in #1185 — that one degrades over hundreds of iterations, whereas yours crashes on the very first backward pass regardless of settings. The synchronize+clear_cache workaround from #1185 won't help here.

Worth checking if the issue is specific to applegpu_g17s (M5 Max) GPU command buffer limits rather than actual memory pressure — the error is `kIOGPUCommandBufferCallbackErrorOutOfMemory` which can be a GPU resource limit, not a RAM limit. The fact that system memory usage is low at crash time supports that theory.

**Comment by ivanfioravanti** [comment]
Same issue here.

**Comment by Goekdeniz-Guelmez** [comment]
@ivanfioravanti @lawcontinue @sivavenkatay 
it can be due to the hybrid linear architecture of qwen3.5 models beacuse of the nature of the hidden states and the recurrent/linear scans. Standard Transformers store activations for every token in the sequence to compute the gradient. hybrid-linear models use a recurrent state. The Issue is during training, to perform backpropagation through time (BPTT) or the parallel scan operation, the model must store the intermediate states for every single token in the sequence. While Transformers have O(n2) memory for the attention matrix, Hybrid models have O(n×d) memory for states. And because these states have high dimensionality and are stored as dense tensors that may not be as efficiently compressed by the  graph compiler as standard matrix multiplications are. Thats why training works for Qwen3 (transformer) but not for Qwen3.5 (hybrid).

**Comment by lawcontinue** [comment]
Can confirm from the implementation side — we hit the same wall building DeltaNet layers for Qwen3.6-27B on MLX.

The core issue Goekdeniz identified (recurrent states not fitting MLX's graph compiler) matches what we see: parallel scan operations produce dense intermediate tensors that MLX can't optimize the same way it does standard attention matmuls. Our cumsum-based parallel scan works fine for inference (~2000 tok/s per layer, 5120 dim), but backprop through those scans would require storing the full scan trajectory — which is exactly the O(n×d) memory explosion described above.

One thing we noticed: the crash happens specifically when MLX tries to build the backward graph for the scan op, not during the forward pass. So a potential workaround might be checkpointing the scan (recompute during backward instead of storing), similar to gradient checkpointing in PyTorch — though this hasn't been implemented in MLX yet.

Relevant context: we implemented parallel DeltaNet layers in pure MLX (cumsum optimization, correctness verified < 1e-6 vs sequential), so if anyone wants to test hybrid training fixes, the infrastructure is there.

**Comment by airowe** [comment]
This is happening to me on an M1 Pro. Using the same models on ollama worked fine.

**Comment by Goekdeniz-Guelmez** [comment]
Hey everyone, I suggest you guys to look into my package, it already has efficient long context linear training. the API usage is almost the same as mlx-lm with minor additions and changes. https://github.com/Goekdeniz-Guelmez/mlx-lm-lora this also has a ton of more algorithms.

**Comment by brunodantas** [comment]
Not reproducible on base M5 (MacBook Air, 16 GB, macOS 26.6.1).

Ran the exact command from the description (plus `--iters 5`) with mlx-lm 0.31.3, on both mlx 0.32.2 and mlx 0.31.2 pinned to match the report. Both complete the backward pass and train normally: `Iter 5: Train loss 2.690, Peak mem 5.605 GB`.

So this looks specific to the M5 Max (`applegpu_g17s`) or its memory config rather than the M5 generation or Qwen3.5's architecture.

**Comment by sharmadivyanshu** [comment]
Another data point — different chip generation and a smaller model, plus a same-size control.

**Environment:** Apple M3 Pro, 18GB unified memory, macOS 26.5, mlx 0.32.1, mlx-lm 0.31.3, Python 3.14.

**Failing:** `mlx-community/Qwen3.5-4B-4bit` — same symptom as the OP: validation forward pass completes, then Metal OOM at the first backward pass, regardless of configuration (including `--grad-checkpoint`). Generation with the same model works fine.

**Control on the same machine:** `mlx-community/Qwen3-4B-4bit` trains cleanly with identical settings and data:

```
mlx_lm.lora --model mlx-community/Qwen3-4B-4bit --train \
  --data <1.6k NL->PromQL pairs> --iters 800 --batch-size 4 \
  --num-layers 16 --mask-prompt --grad-checkpoint
```

Peak memory 7.9GB (3.8GB on a shorter-prompt variant of the dataset) — nowhere near the 18GB budget. Same size class, same 4-bit quant, same recipe: Qwen3 trains, Qwen3.5 OOMs at the first backward. On this machine it is not a capacity problem.

Given the non-repro report on the base M5 above: this adds M3 Pro to the failing side (alongside M5 Max and M1 Pro), and 4B alongside 9B — whatever the trigger is, it doesn't look specific to one chip generation or model size.
